Tests/SwiftSyntaxMacrosTest/MacroSystemTests.swift

```diff
@@ -145,6 +145,125 @@ public struct FileIDMacro: ExpressionMacro {
   }
 }
 
+extension PatternBindingSyntax {
+  /// When the variable is declaring a single binding, produce the name of
+  /// that binding.
+  fileprivate var singleBindingName: String? {
+    if let identifierPattern = pattern.as(IdentifierPatternSyntax.self) {
+      return identifierPattern.identifier.text
+    }
+
+    return nil
+  }
+}
+
+extension SyntaxProtocol {
+  /// Form a function name.
+  private func formFunctionName(
+    _ baseName: String, _ parameters: ParameterClauseSyntax?,
+    isSubscript: Bool = false
+  ) -> String {
+    let argumentNames: [String] = parameters?.parameterList.map { param in
+      let argumentLabelText = param.argumentName?.text ?? "_"
+      return argumentLabelText + ":"
+    } ?? []
+
+    return "\(baseName)(\(argumentNames.joined(separator: "")))"
+  }
+
+  /// Form the #function name for the given node.
+  fileprivate func functionName<Context: MacroExpansionContext>(
+    in context: Context
+  ) -> String? {
+    // Declarations with parameters.
+    // FIXME: Can we abstract over these?
+    if let function = self.as(FunctionDeclSyntax.self) {
+      return formFunctionName(
+        function.identifier.text, function.signature.input
+      )
+    }
+
+    if let initializer = self.as(InitializerDeclSyntax.self) {
+      return formFunctionName("init", initializer.signature.input)
+    }
+
+    if let subscriptDecl = self.as(SubscriptDeclSyntax.self) {
+      return formFunctionName(
+        "subscript", subscriptDecl.indices, isSubscript: true
+      )
+    }
+
+    if let enumCase = self.as(EnumCaseElementSyntax.self) {
+      guard let associatedValue = enumCase.associatedValue else {
+        return enumCase.identifier.text
+      }
+
+      let argumentNames = associatedValue.parameterList.map { param in
+        guard let firstName = param.firstName else {
+          return "_:"
+        }
+
+        return firstName.text + ":"
+      }.joined()
+
+      return "\(enumCase.identifier.text)(\(argumentNames))"
+    }
+
+    // Accessors use their enclosing context, i.e., a subscript or pattern
+    // binding.
+    if self.is(AccessorDeclSyntax.self) {
+      guard let parentContext = context.parentContext(of: self) else {
+        return nil
+      }
+
+      return parentContext.functionName(in: context)
+    }
+
+    // All declarations with identifiers.
+    if let identified = self.asProtocol(IdentifiedDeclSyntax.self) {
+      return identified.identifier.text
+    }
+
+    // Extensions
+    if let extensionDecl = self.as(ExtensionDeclSyntax.self) {
+      // FIXME: It would be nice to be able to switch on type syntax...
```

Comment:
You could switch over `extensionDecl.kind` and then get the meta type using `SyntaxKind.syntaxNodeType`. You could also use that above.